### PR TITLE
Get tensorflow, jax, and pytorch working on TPU1VM

### PR DESCRIPTION
**Pull request by djherbis**
http://b/213335159

### CODE REVIEW

**Review by Philmod — APPROVED** [review]

File: tpu/config.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+BASE_IMAGE_TAG=v108
```
Comment (djherbis) [inline tpu/config.txt:1]: This is the latest CPU build I tested this against that I believe works. Will bump it eventually.

**Comment by djherbis** [comment]
@rosbo Sending to you since I added jenkins stuff (this builds locally on my machine at least when I run the ./tpu/build script.

File: tpu/Dockerfile [hunk 2]
```diff
@@ -6,8 +6,19 @@ FROM gcr.io/cloud-tpu-v2-images/libtpu:${LIBTPU_IMAGE_TAG} as libtpu
...
 
 COPY --from=tensorflow_whl /tmp/tensorflow_pkg/tensorflow*.whl /tmp/tensorflow_pkg/
 RUN pip install /tmp/tensorflow_pkg/tensorflow*.whl && \
-    rm -rf /tmp/tensorflow_pkg
\ No newline at end of file
+    rm -rf /tmp/tensorflow_pkg
+
+# https://cloud.google.com/tpu/docs/pytorch-xla-ug-tpu-vm#changing_pytorch_version
+RUN pip uninstall -y torch
+RUN pip uninstall -y torch_xla
```
Comment (rosbo) [inline tpu/Dockerfile:19]: I don't think `torch_xla` is installed on the base image.

File: tpu/Dockerfile [hunk 3]
```diff
@@ -6,8 +6,19 @@ FROM gcr.io/cloud-tpu-v2-images/libtpu:${LIBTPU_IMAGE_TAG} as libtpu
...
 COPY --from=libtpu /libtpu.so /lib
 
 COPY --from=tensorflow_whl /tmp/tensorflow_pkg/tensorflow*.whl /tmp/tensorflow_pkg/
 RUN pip install /tmp/tensorflow_pkg/tensorflow*.whl && \
-    rm -rf /tmp/tensorflow_pkg
\ No newline at end of file
+    rm -rf /tmp/tensorflow_pkg
+
+# https://cloud.google.com/tpu/docs/pytorch-xla-ug-tpu-vm#changing_pytorch_version
+RUN pip uninstall -y torch
```
Comment (rosbo) [inline tpu/Dockerfile:18]: Make sure to clean the layer at the end of each `RUN` command otherwise the layer will include the compressed packages which makes the layer larger and increases the image size.

https://github.com/Kaggle/docker-python/blob/389cecd4683774c7ce33d39667dd788742e77816/Dockerfile.tmpl#L104-L107

I would recommend batching the RUN command and running a process to clean layer at the end.

File: Jenkinsfile [hunk 4]
```diff
@@ -161,6 +188,7 @@ pipeline {
 
           gcloud container images add-tag gcr.io/kaggle-images/python:${PRETEST_TAG} gcr.io/kaggle-images/python:${STAGING_TAG}
           gcloud container images add-tag gcr.io/kaggle-private-byod/python:${PRETEST_TAG} gcr.io/kaggle-private-byod/python:${STAGING_TAG}
+          gcloud container images add-tag gcr.io/kaggle-private-byod/python-tpuvm:${PRETEST_TAG} gcr.io/kaggle-private-byod/python-tpuvm:${STAGING_TAG}
```
Comment (rosbo) [inline Jenkinsfile:191]: The `staging` label should be added after the tests are passing. Should you run the tests here?

File: Jenkinsfile [hunk 5]
```diff
@@ -150,7 +150,34 @@ pipeline {
               }
             }
           }
-        } 
+        }
+        stage('TPU VM') {
```
Comment (rosbo) [inline Jenkinsfile:154]: Given that this image is built against a fixed version of the CPU image (here, v108), I don't think we should rebuild it all the time. Or is it the goal to not pin it to v108 eventually?

File: Jenkinsfile [hunk 6]
```diff
@@ -150,7 +150,34 @@ pipeline {
...
+          stages {
+            stage('Build Tensorflow TPU Image') {
+              options {
+                timeout(time: 120, unit: 'MINUTES')
```
Comment (rosbo) [inline Jenkinsfile:158]: How long does this step take?

File: Jenkinsfile (hunk 4 above)
Comment (djherbis) [reply Jenkinsfile:191]: I didn't have any tests setup yet, in part because we don't have a TPU attached in jenkins to test against, and otherwise it has basically the same libs as the CPU image (which gets tested already).

I would like to figure out a testing plan eventually, but figured I'd just start by matching the existing setup as much as possible.

File: Jenkinsfile (hunk 6 above)
Comment (djherbis) [reply Jenkinsfile:158]: I've never run it on Jenkins so I guess we don't know yet for sure. Building the tensorflow image takes considerable time though :(

File: tpu/Dockerfile (hunk 3 above)
Comment (djherbis) [reply tpu/Dockerfile]: Done! Thanks.

File: tpu/Dockerfile (hunk 2 above)
Comment (djherbis) [reply tpu/Dockerfile]: Done. Thanks!

File: Jenkinsfile (hunk 4 above)
Comment (rosbo) [reply Jenkinsfile:191]: Makes sense. Thanks for the clarification.

File: Jenkinsfile (hunk 6 above)
Comment (rosbo) [reply Jenkinsfile:158]: How long does building the TPU VM image takes? (without building TensorFlow)?

I am concern with adding an extra 2 hours to every build. Having this build complete in the shortest amount of time is very valuable for me to keep the maintenance cost low and iterate fast.

The TensorFlow Image probably doesn't need to be rebuilt every time. It should only be rebuilt when the base image or the TensorFlow image is updated. We could use a similar approach than what I use for building lightgbm and torch from source for the GPU image. I basically build the packages in parallel and the pipeline only rebuild the image if the base image or the package version is updated: 

- logic for parallel build: https://github.com/Kaggle/docker-python/blob/389cecd4683774c7ce33d39667dd788742e77816/Jenkinsfile#L24-L59
- logic for conditional build: https://github.com/Kaggle/docker-python/blob/389cecd4683774c7ce33d39667dd788742e77816/packages/build_package#L118-L131

File: tpu/config.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
...
+LIBTPU_IMAGE_TAG=libtpu_1.1.0_RC00
+TENSORFLOW_WHL_IMAGE_TAG=local
```
Comment (rosbo) [inline tpu/config.txt:3]: Should this be the image build by this pipeline?

File: Jenkinsfile (hunk 6 above)
Comment (djherbis) [reply Jenkinsfile]: Thanks! I think I've adjusted it to fit this (merged it into the "build_packages" setup).

File: tpu/config.txt (hunk 7 above)
Comment (djherbis) [reply tpu/config.txt]: Yes, cleaned this up.

File: tpu/Dockerfile [hunk 8]
```diff
@@ -1,13 +1,26 @@
...
 COPY --from=tensorflow_whl /tmp/tensorflow_pkg/tensorflow*.whl /tmp/tensorflow_pkg/
 RUN pip install /tmp/tensorflow_pkg/tensorflow*.whl && \
-    rm -rf /tmp/tensorflow_pkg
\ No newline at end of file
+    rm -rf /tmp/tensorflow_pkg && \
+    /tmp/clean-layer.sh
+
+# https://cloud.google.com/tpu/docs/pytorch-xla-ug-tpu-vm#changing_pytorch_version
+RUN pip uninstall -y torch && \
+    pip install torch==1.10 && \
```
Comment (rosbo) [inline tpu/Dockerfile:20]: You can use `TORCH_VERSION` if you want. `main` is currently at `1.11`: https://github.com/Kaggle/docker-python/blob/389cecd4683774c7ce33d39667dd788742e77816/Dockerfile.tmpl#L6

File: tpu/config.txt (hunk 1 above)
Comment (rosbo) [reply tpu/config.txt:1]: Can you add a TODO explaining this and the end goal of setting this to `ci-pretest`.

**Review by rosbo — APPROVED** [review]

File: tpu/Dockerfile (hunk 8 above)
Comment (djherbis) [reply tpu/Dockerfile]: Added, for now, I replicated this arg into tpu/config.txt since it may drift, but the eventual plan would be to reuse it when we move to ci-pretest.
